pkg/recommendedversion/recommended_version.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2024 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package recommendedversion is used to check for
+// the currently recommended versions of the Tanzu CLI
+// and inform the user if they are using an outdated version.
+package recommendedversion
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+
+	"github.com/vmware-tanzu/tanzu-cli/pkg/buildinfo"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/centralconfig"
+	cliconfig "github.com/vmware-tanzu/tanzu-cli/pkg/config"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/constants"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/datastore"
+	"github.com/vmware-tanzu/tanzu-cli/pkg/utils"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/config"
+	"github.com/vmware-tanzu/tanzu-plugin-runtime/log"
+)
+
+// dataStoreLastVersionCheckKey is the data store key used to store the last
+// time the version check was done
+const (
+	centralConfigRecommendedVersionsKey = "cli.core.cli_recommended_versions"
```

Comment:
After thinking a bit more and the way the API is designed to get the Key/Value, I feel this might cause potential confusion to the user. (I should have raised this as part of #700 but will put my thoughts here as the other one is merged).

I see that we are using key names as `cli.core.cli_recommended_versions` in the central configuration. Considering this is a yaml file that can grow in the future, should we be using a yaml way of defining this instead of just plain `key:value`?

For example:
```
cli:
  core:
    cli_recommended_versions: 
      - v2.1.0-alpha.2
      - v2.0.2
      - v1.5.0-beta.0
      - v1.4.4
      - v1.3.3
      - v1.2.2
      - v1.1.1
      - v0.90.0
```

So when retrieving the `cli_recommended_version` information we can invoke the API as `GetCentralConfigEntry("cli.core.cli_recommended_versions")`? 
This can be done using https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured package in a generalized way as the package allows different APIs to retrieve a specific nested resource from `map[string]interface{}`.

_Just adding my thoughts here so that we can think which approach is better and discuss further if needed._

Reply:
My thought was that each feature using the central config would have the knowledge of the data it needs.
So, the API allows key:value only as `value` is very feature-specific and it is the feature code that will know how to decode it.  I didn't feel the central config API should be into too much flexibility but instead leave the feature deal with that.  For example, if a feature wants to save a yaml text as the `value` then it can fetch the whole `value` using this API and then use `yaml.Unmarshal`.  

That way, the feature (or later on plugins) could use json, csv, or any format they want, not only yaml.

But I'll admit I'm having trouble seeing if the API would be annoying to use.
If you have examples in mind that would help me better understand.

Reply:
The main concern I have is that, if the configuration grows in the future, it might be hard to maintain and might not be easily readable.

>So, the API allows key:value only as value is very feature-specific and it is the feature code that will know how to decode it. I didn't feel the central config API should be into too much flexibility but instead leave the feature deal with that. For example, if a feature wants to save a yaml text as the value then it can fetch the whole value using this API and then use yaml.Unmarshal.

I was thinking that implementing a fine-grained (specific) API to get the required information might be sufficient to address this. Meaning instead of providing the `GetCentralConfigEntry` API as part of the `CentralConfig` interface, we can implement `GetRecommenededVersions` directly. 

I will spend some time rethinking if there are any other limitations as I am not able to think of anything besides the readability of this (possibly a huge) file. But as the user does not need to read that file, we can get away with the way it is defined now.

Reply:
Thanks for explaining this. I got confused earlier but the way it is implemented should be fine. So, this type of key in only for a root level key, and we can define anything below that key.

In that case, is it safe to say we can define the above configuration as:
```
cli.core.cli_recommended_versions: 
      - v2.1.0-alpha.2
      - v2.0.2
      - v1.5.0-beta.0 
      - v1.4.4
      - v1.3.3  # Some comment
      - v1.2.2
      - v1.1.1.  # This structure will allow us to add any specific comment to a version
      - v0.90.0
```

Basically, instead of using `string` for the version, can we use an array?

Reply:
Good idea. I’ll read the yank-array value with the API then unmarshal it to get a real array. 

I’ll update the PR

Reply:
yank-array -> yaml-array

no joke, auto-correct changed it on me not once but three times!

Reply:
Working with @anujc25 we improved the API.  `GetCentralConfigEntry(key string, out interface{})` now also takes an `out` pointer and the method will unmarshall the value into this `out` variable respecting its type.  This allows the caller to get a proper Go type that match the value being read from the central config.